repoze/lru/__init__.py

```diff
@@ -270,37 +327,50 @@ class lru_cache(object):
     timeout parameter specifies after how many seconds a cached entry should
     be considered invalid.
     """
-    def __init__(self, maxsize, cache=None, timeout=None): # cache is an arg to serve tests
+
+    def __init__(self, maxsize, cache=None, timeout=None):  # cache is an arg to serve tests
         if cache is None:
-            if timeout is None:
+            if maxsize is None:
+                cache = UnboundedCache()
+            elif timeout is None:
                 cache = LRUCache(maxsize)
             else:
                 cache = ExpiringLRUCache(maxsize, default_timeout=timeout)
         self.cache = cache
 
-    def __call__(self, f):
+    def __call__(self, func):
         cache = self.cache
         marker = _MARKER
-        def lru_cached(*arg):
-            val = cache.get(arg, marker)
-            if val is marker:
-                val = f(*arg)
-                cache.put(arg, val)
-            return val
+
+        def cached_wrapper(*args, **kwargs):
+            try:
+                key = (args, frozenset(kwargs.items())) if kwargs else args
+                val = cache.get(key, marker)
+                if val is marker:
+                    val = func(*args, **kwargs)
+                    cache.put(key, val)
+                return val
+            except TypeError:  # unhashable types in args or kwargs
+                if _WARN:
+                    warn("method call could not be cached, *args and/or **kwargs contain unhashable types")
```

Comment:
true, I've modified the code to support this behaviour if explicitly enabled (by setting the `ignore_unhashable_args` arg). By default, the exception is raised as expected.